File: source/python/atomic_counter.py
import threading

counter = 0
lock = threading.Lock()
num_tasks = 10
iterations = 100

def increment_worker():
    global counter
    for _ in range(iterations):
        with lock:
            counter += 1

threads = [threading.Thread(target=increment_worker) for _ in range(num_tasks)]
for t in threads:
    t.start()
for t in threads:
    t.join()

print(f"final: {counter}")

File: source/rust/atomic_counter.rs
use std::sync::{Arc, atomic::{AtomicI32, Ordering}};
use std::thread;

fn main() {
    let counter = Arc::new(AtomicI32::new(0));
    let num_tasks = 10;
    let iterations = 100;

    let mut handles = Vec::new();
    for _ in 0..num_tasks {
        let c = Arc::clone(&counter);
        handles.push(thread::spawn(move || {
            for _ in 0..iterations {
                c.fetch_add(1, Ordering::SeqCst);
            }
        }));
    }
    for h in handles { h.join().unwrap(); }
    println!("final: {}", counter.load(Ordering::SeqCst));
}
